File: 30_days_of_code/day12/day12.cpp
#include <iostream>
#include <string>
#include <vector>

class Person {
protected:
  std::string firstName;
  std::string lastName;
  int id;

public:
  Person(std::string firstName, std::string lastName, int identification)
  {
    this->firstName = firstName;
    this->lastName = lastName;
    this->id = identification;
  }
  void printPerson()
  {
    std::cout << "Name: " << lastName << ", " << firstName << "\nID: " << id
              << std::endl;
  }
};

class Student : public Person {
private:
  std::vector<int> testScores;
  int numOfTestScores = 0;
  int sum = 0;
  double average = 0;

public:
  Student(std::string firstName, std::string LastName, int id,
          std::vector<int> scores)
      : Person(firstName, LastName, id)
  {
    this->testScores = scores;
    this->numOfTestScores = scores.size();
  }

  char calculate()
  {
    char outputChar;
    for (unsigned int i = 0; i < numOfTestScores; ++i) {
      sum += testScores[i];
    }

    average = sum / numOfTestScores;
    if (average >= 90 && average <= 100)
      outputChar = 'O';

    else if (average >= 80 && average < 90)
      outputChar = 'E';

    else if (average >= 70 && average < 80)
      outputChar = 'A';

    else if (average >= 55 && average < 70)
      outputChar = 'P';

    else if (average >= 40 && average < 55)
      outputChar = 'D';

    else if (average < 40)
      outputChar = 'T';

    return outputChar;
  }
};

int main()
{
  std::string firstName;
  std::string lastName;
  int id = 0;
  int numScores = 0;
  std::cin >> firstName >> lastName >> id >> numScores;

  std::vector<int> scores;
  for (int i = 0; i < numScores; i++) {
    int tmpScore;
    std::cin >> tmpScore;
    scores.push_back(tmpScore);
  }

  Student *s = new Student(firstName, lastName, id, scores);
  s->printPerson();
  std::cout << "Grade: " << s->calculate() << std::endl;

  return 0;
}

File: 30_days_of_code/day12/day12.py
class Person:
    def __init__(self, firstName, lastName, idNumber):
        self.firstName = firstName
        self.lastName = lastName
        self.idNumber = idNumber

    def printPerson(self):
        print("Name:", self.lastName + ",", self.firstName)
        print("ID:", self.idNumber)


class Student(Person):
    #   Class Constructor
    #
    #   Parameters:
    #   firstName - A string denoting the Person's first name.
    #   lastName - A string denoting the Person's last name.
    #   id - An integer denoting the Person's ID number.
    #   scores - An array of integers denoting the Person's test scores
    def __init__(self, firstName, lastName, idNumber, testScores):
        super().__init__(firstName, lastName, idNumber)
        self.scores = testScores

    #   Function Name: calculate
    #   Return: A character denoting the grade
    def calculate(self):
        average = 0.0
        for testScore in self.scores:
            average += testScore

        outputChar = ' '

        average = average / len(self.scores)

        if(average >= 90 and average <= 100):
            outputChar = 'O'

        elif(average >= 80 and average < 90):
            outputChar = 'E'

        elif(average >= 70 and average < 80):
            outputChar = 'A'

        elif(average >= 55 and average < 70):
            outputChar = 'P'

        elif(average >= 40 and average < 55):
            outputChar = 'D'

        elif(average < 40):
            outputChar = 'T'

        return outputChar


line = input().split()
firstName = line[0]
lastName = line[1]
idNum = line[2]
numScores = int(input())
scores = list(map(int, input().split()))
s = Student(firstName, lastName, idNum, scores)
s.printPerson()
print("Grade:", s.calculate())
